### PR TITLE
Fix/56 remove hasattr getattr etc

### LINKED ISSUE #56: Remove hasattr, getattr, etc.
Remove hasattr, getattr, globals(),  etc. and instead relay on static type-hinting. The current setup is prone to bugs.

### PR DESCRIPTION
**Pull request by mortendaehli**
# Pull Request

### WHAT

Fix static code issues related to hasattr and getattr, and resolve some other general issues.

### Review:

Need to resolve all my commented code issues below. These are either bugs or very prone to create a bug in the future.

## Checklist:

- [x] :tada: This PR closes #56 .

### CODE REVIEW

File: src/oogeso/core/devices/gasheater.py [hunk 1]
```diff
@@ -24,7 +23,7 @@ def __init__(
 
     def _rules(self, pyomo_model: pyo.Model, t: int) -> Union[pyo.Expression, pyo.Constraint.Skip]:
         dev = self.id
-        param_dev = self.params  # Fixme: Attribute params is missing
+        param_dev = self.params  # Fixme: Attribute params is missing.
```
Comment (mortendaehli) [inline src/oogeso/core/devices/gasheater.py:26]: These are missing. self.params are never set. What is the idea here?

File: src/oogeso/core/devices/separator.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def _rule_separator(self, pyomo_model: pyo.Model, t: int, i: int) -> Union[pyo.E
         dev_data = self.dev_data
         node = dev_data.node_id
         node_obj: NetworkNode = self.optimiser.all_nodes[node]  # Fixme: Unresolved reference
-        wellstream_prop: dto.CarrierWellstreamData = self.all_carriers["wellstream"]  # Fixme: Unresolved reference
+        wellstream_prop: dto.CarrierWellStreamData = self.all_carriers["wellstream"]  # Fixme: Unresolved reference
```
Comment (mortendaehli) [inline src/oogeso/core/devices/separator.py:32]: We need to resolve self.optimiser.all_nodes and self.all_varriers

File: src/oogeso/core/devices/storage.py [hunk 3]
```diff
@@ -247,7 +247,7 @@ def compute_costForDepletedStorage_alt2(self, pyomo_model: pyo.Model, timesteps:
         dev_data = self.dev_data
         # E_target = dev_data.E_max
         E_target = pyomo_model.paramDeviceEnergyTarget[self.id]
-        t_end = timesteps.last()
+        t_end = timesteps[-1]
```
Comment (mortendaehli) [inline src/oogeso/core/devices/storage.py:250]: Could this be anything other than a List[int]?

File: src/oogeso/core/networks/edge.py [hunk 4]
```diff
@@ -99,9 +101,12 @@ def _loss_function_constraint(self, i: int, pyomo_model: pyo.Model, piecewise_re
...
         # Losses on cables are: P_loss = R/V^2 * P^2, i.e. quadratic function of power flow
         # Losses in transformers are: P_loss = ...
-        lookup_table = self.edge_data.power_loss_function
-        pw_x = lookup_table[0]
-        pw_y = lookup_table[1]
+        if hasattr(self.edge_data, "power_loss_function"):
+            lookup_table = self.edge_data.power_loss_function  # Fixme: What do we do if there are no loss functions?
+            pw_x = lookup_table[0]
+            pw_y = lookup_table[1]
+        else:  # Fixme: Only certain Edges expects loss function... Should we default power loss function to 0, 0?
```
Comment (mortendaehli) [inline src/oogeso/core/networks/edge.py:108]: Could this happen? How do we want to handle this? It seems like the code expect a certain type of edge that has a loss function. But not all of them does? Or should we implement this is in the base class?

File: src/oogeso/core/networks/edge.py [hunk 5]
```diff
@@ -128,3 +133,11 @@ def _loss_function_constraint(self, i: int, pyomo_model: pyo.Model, piecewise_re
             f_rule=pw_y,
         )
         return constr_penalty
+
+
+class FluidEdge(Edge):
```
Comment (mortendaehli) [inline src/oogeso/core/networks/edge.py:138]: Is this usefull somehow? Ref. expecting loss function above. Maybe we should have a family of Edges inheriting from Edge?

File: src/oogeso/core/networks/electricity.py [hunk 6]
```diff
@@ -16,7 +17,7 @@ class El(Network):
     def __init__(
         self,
         carrier_data: dto.CarrierElData,
-        edges: Dict[str, dto.EdgeElData],
+        edges: Dict[str, ElEdge],  # Fixme: Earlier we passed the DTO, but the code expects the an Edge core object.
```
Comment (mortendaehli) [inline src/oogeso/core/networks/electricity.py:20]: Should we rename the class to ElNetwork?

File: src/oogeso/core/networks/electricity.py [hunk 7]
```diff
@@ -30,7 +31,9 @@ def define_constraints(self, pyomo_model: pyo.Model) -> None:
         if self.carrier_data.powerflow_method == "dc-pf":
             logger.warning("TODO: code for electric powerflow calculations need improvement (pu conversion)")
             nodelist = list(pyomo_model.setNode)  # self.all_nodes.keys()
-            edgelist_el = {edge_id: asdict(edge.edge_data) for edge_id, edge in self.edges.items()}
+            edgelist_el = {
```
Comment (mortendaehli) [inline src/oogeso/core/networks/electricity.py:34]: Need to resolve this potential bug.

File: src/oogeso/core/networks/pipelines.py [hunk 8]
```diff
@@ -47,15 +50,18 @@ def _rulePipelineFlow(self, model: pyo.Model, edge, t: int) -> Union[pyo.Express
...
-            if print_log:
-                logger.debug("{},{}: {} parallel pipes".format(edge, t, num_pipes))
-            Q = Q / num_pipes
+        if hasattr(edge_data, "num_pipes"):
+            if edge_data.num_pipes is not None:
+                num_pipes = edge_data.num_pipes
+                if print_log:
+                    logger.debug("{},{}: {} parallel pipes".format(edge, t, num_pipes))
+                Q = Q / num_pipes
+        else:
```
Comment (mortendaehli) [inline src/oogeso/core/networks/pipelines.py:59]: Is this as expected?

File: src/oogeso/core/networks/pipelines.py [hunk 9]
```diff
@@ -15,6 +15,9 @@
 
 
 class Fluid(Network):
+    carrier_data: dto.CarrierFluidData
```
Comment (mortendaehli) [inline src/oogeso/core/networks/pipelines.py:18]: Shoulf we name the class FluidNetwork?

File: src/oogeso/core/networks/pipelines.py [hunk 10]
```diff
@@ -118,9 +128,7 @@ def compute_edge_pressuredrop(self, edge: "Edge", p1, Q, method=None, linear=Fal
             if linear & (p0_from == p0_to):
                 method = None
                 if print_log:
-                    logger.debug(
-                        ("{}-{}: Pipe without pressure drop" " ({} / {} MPa)").format(n_from, n_to, p0_from, p0_to)
-                    )
+                    logger.debug(f"{n_from}-{n_to}: Pipe without pressure drop" " ({p0_from} / {p0_to} MPa)")
```
Comment (mortendaehli) [inline src/oogeso/core/networks/pipelines.py:131]: Can we remove this print_log thing?

File: src/oogeso/core/optimiser.py [hunk 11]
```diff
@@ -351,7 +354,8 @@ def _specify_constraints(self):
         else:
             logger.debug("No emission_intensity_max limit specified")
         # 4.3 electrical reserve margin:
-        el_parameters = self.all_networks["el"].carrier_data
+        # Fixme: This is prone to errors, and depends on the class El(Network).__name__ not changing.
```
Comment (mortendaehli) [inline src/oogeso/core/optimiser.py:357]: Need to resolve this..

File: src/oogeso/core/optimiser.py [hunk 12]
```diff
@@ -351,7 +354,8 @@ def _specify_constraints(self):
...
+        el_parameters: dto.CarrierElData = self.all_networks["el"].carrier_data  # Fixme: Static type error.
```
Comment (mortendaehli) [inline src/oogeso/core/optimiser.py:358]: And this

File: src/oogeso/core/optimiser.py [hunk 13]
```diff
@@ -531,10 +536,16 @@ def _rule_el_reserve_margin(self, pyomo_model: pyo.Model, t: int) -> Union[pyo.E
...
-        expr = capacity_unused >= margin
-        return expr
+        if hasattr(network_el.carrier_data, "el_reserve_margin"):
+            margin = network_el.carrier_data.el_reserve_margin
+        else:
+            raise ValueError("network_el.carrier_data does not have the attribute el_reserve_margin")
+        if hasattr(network_el, "compute_el_reserve") and callable(getattr(network_el, "compute_el_reserve")):
+            capacity_unused = network_el.compute_el_reserve(pyomo_model=pyomo_model, t=t, all_devices=self.all_devices)
+        else:
+            raise ValueError("network_el does not have the function compute_el_reserve")
```
Comment (mortendaehli) [inline src/oogeso/core/optimiser.py:546]: How do we want to handle this? Should we expect this to happen?

File: src/oogeso/core/optimiser.py [hunk 14]
```diff
@@ -531,10 +536,16 @@ def _rule_el_reserve_margin(self, pyomo_model: pyo.Model, t: int) -> Union[pyo.E
...
 
         # Fixme: Error in type hint.
-        margin = network_el.carrier_data.el_reserve_margin
-        capacity_unused = network_el.compute_el_reserve(pyomo_model=pyomo_model, t=t, all_devices=self.all_devices)
-        expr = capacity_unused >= margin
-        return expr
+        if hasattr(network_el.carrier_data, "el_reserve_margin"):
+            margin = network_el.carrier_data.el_reserve_margin
+        else:
+            raise ValueError("network_el.carrier_data does not have the attribute el_reserve_margin")
```
Comment (mortendaehli) [inline src/oogeso/core/optimiser.py:542]: How do we want to handle this? Should we expect this to happen?

File: src/oogeso/dto/oogeso_input_data_objects_multienergy.py [hunk 15]
```diff
@@ -97,17 +102,22 @@ class DeviceGasTurbineData(dto.DeviceData):
 
 
 @dataclass
-class DeviceHeatPumpData(dto.DeviceData):
+class DevicePumpData(dto.DeviceData):
```
Comment (mortendaehli) [inline src/oogeso/dto/oogeso_input_data_objects_multienergy.py:105]: Do we need one class for each Pump Device?

File: src/oogeso/dto/oogeso_input_data_objects_multienergy.py [hunk 16]
```diff
@@ -204,40 +214,39 @@ class CarrierHydrogenData(dto.CarrierData):
...
-    co2_content: float  # kg/Sm3 - see SSB 2016 report -> 2.34 kg/Sm3
-    G_gravity: float  # 0.6
-    Pb_basepressure_MPa: float  # MPa -> 0.101 # MPa
-    R_individual_gas_constant: float  # J/(kg K) -> 500 J/kgK
-    Tb_basetemp_K: float  # K -> 288 K = 15 degC
-    Z_compressibility: float  # 0.9
-    energy_value: float  # MJ/Sm3 (calorific value) -> 40 MJ/Sm3
-    k_heat_capacity_ratio: float  # 1.27
-    rho_density: float  # kg/m3 -> 0.84 kg/m3
+class CarrierFluidData(dto.CarrierData):
```
[the remark this answers is not in the document]
Comment (mortendaehli) [reply src/oogeso/dto/oogeso_input_data_objects_multienergy.py:217]: Do we need one for each Fluid class?

**Review by frodehk — APPROVED** [review]

File: src/oogeso/core/devices/separator.py (hunk 2 above)
Comment (HaraldGSvendsen) [reply src/oogeso/core/devices/separator.py:32]: also fixed in pull request test: Unit tests for pyomo model device constraints #74

File: src/oogeso/core/devices/storage.py (hunk 3 above)
Comment (HaraldGSvendsen) [reply src/oogeso/core/devices/storage.py:250]: It's a pyo.Set (pyomo_model.setHorizon), so I think it's the type hint that is wrong.  But last() is probably just doing the same as [-1]

File: src/oogeso/core/networks/pipelines.py (hunk 10 above)
Comment (HaraldGSvendsen) [reply src/oogeso/core/networks/pipelines.py:131]: yes

File: src/oogeso/core/networks/edge.py (hunk 5 above)
Comment (mortendaehli) [reply src/oogeso/core/networks/edge.py:138]: Investigate further...

File: src/oogeso/core/networks/electricity.py (hunk 7 above)
Comment (mortendaehli) [reply src/oogeso/core/networks/electricity.py:34]: raise NotImplementedError

File: src/oogeso/core/networks/pipelines.py (hunk 8 above)
Comment (mortendaehli) [reply src/oogeso/core/networks/pipelines.py:59]: Keep as is. Set FluidEdgeData.num_pipes... to 1 as default?

File: src/oogeso/core/optimiser.py (hunk 12 above)
Comment (mortendaehli) [reply src/oogeso/core/optimiser.py:358]: Need to change if we rename El(Network)? See if we can find a better solution than this dictionary.

File: src/oogeso/core/simulator.py [hunk 17]
```diff
@@ -286,7 +286,7 @@ def _save_optimisation_result(
         if return_all or "el_reserve" in return_variables:
             df_reserve = pd.Series(dtype=float64, index=range(timestep, timestep + timelimit))
             for t in range(timelimit):
-                rescap = pyo.value(
+                rescap = pyo.value(  # Fixme: This reference to compute_el_reserve may be incorrect.
```
[the remark this answers is not in the document]
Comment (mortendaehli) [reply src/oogeso/core/simulator.py:289]: Seems to be correct
